sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -1376,111 +1361,14 @@ object GpuCast {
-            isTimestamp.or(nextIsTimeStamp)
-          }
-        }
-      }
-
-      val isValidTimestamp = withResource(isCudfMatch) { _ =>
-        val prog = new RegexProgram(TIMESTAMP_REGEX_FULL, CaptureGroups.NON_CAPTURE)
-        withResource(input.matchesRe(prog)) { isRegexMatch =>
-          isCudfMatch.and(isRegexMatch)
-        }
-      }
-
-      // we only need to parse with one of the cuDF formats because the parsing code ignores
-      // the ' ' or 'T' between the date and time components
-      withResource(isValidTimestamp) { _ =>
-        withResource(input.asTimestampMicroseconds(cudfFormat1)) { asDays =>
-          isValidTimestamp.ifElse(asDays, orElse)
-        }
-      }
-    }
-  }
-
-  def castStringToTimestamp(input: ColumnVector, ansiMode: Boolean): ColumnVector = {
-
-    // special timestamps
-    val today = DateUtils.currentDate()
-    val todayStr = new SimpleDateFormat("yyyy-MM-dd")
-        .format(today * DateUtils.ONE_DAY_SECONDS * 1000L)
-
-    var sanitizedInput = input.incRefCount()
-
-    // prepend today's date to timestamp formats without dates
-    sanitizedInput = withResource(sanitizedInput) { _ =>
-      val prog = new RegexProgram(TIMESTAMP_REGEX_NO_DATE)
-      sanitizedInput.stringReplaceWithBackrefs(prog, s"${todayStr}T\\1")
-    }
-
-    withResource(sanitizedInput) { sanitizedInput =>
-      // convert dates that are in valid timestamp formats
-      val converted =
-        convertFullTimestampOr(sanitizedInput,
-          convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_YYYY_MM_DD, "%Y-%m-%d",
-            convertTimestampOr(sanitizedInput, TIMESTAMP_REGEX_YYYY_MM, "%Y-%m",
-              convertTimestampOrNull(sanitizedInput, TIMESTAMP_REGEX_YYYY, "%Y"))))
-
-      // handle special dates like "epoch", "now", etc.
-      val finalResult = closeOnExcept(converted) { tsVector =>
-        DateUtils.fetchSpecialDates(DType.TIMESTAMP_MICROSECONDS) match {
-          case specialDates if specialDates.nonEmpty =>
-            // `tsVector` will be closed in replaceSpecialDates.
-            replaceSpecialDates(sanitizedInput, tsVector, specialDates)
-          case _ =>
-            tsVector
-        }
-      }
-
-      if (ansiMode) {
-        // When ANSI mode is enabled, we need to throw an exception if any values could not be
-        // converted
-        checkResultForAnsiMode(input, finalResult,
-          "One or more values could not be converted to TimestampType")
+  def castStringToTimestamp(input: ColumnView, ansiMode: Boolean, defaultTimeZone: Option[String] = Option.empty[String]): ColumnVector = {
+    val tz = defaultTimeZone.getOrElse("Z")
+    val normalizedTZ = ZoneId.of(tz, ZoneId.SHORT_IDS).normalized().toString
+    withResource(CastStrings.toTimestamp(input, normalizedTZ, ansiMode)) { result =>
+      if (ansiMode && result == null) {
+        throw new RuntimeException("DateTimeException")
       } else {
-        finalResult
+        result.incRefCount()
```

Comment:
This can be simplifed as
```
    closeOnExcept(CastStrings.toTimestamp(input, normalizedTZ, ansiMode)) { result =>
      if (ansiMode && result == null) {
        throw new RuntimeException("DateTimeException")
      }
      result
    }
```